{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Class containing functions to load data needed for HMM Run.\n",
    "Returns all relevant data for HMM in unified format:\n",
    "1) Haplotype Probabilities \n",
    "2) Allele Frequencies \n",
    "3) Genetic Map\n",
    "@ Author: Harald Ringbauer, September 2020\n",
    "\"\"\"\n",
    "\n",
    "import pickle as pickle\n",
    "import numpy as np\n",
    "import os as os\n",
    "import h5py as h5py\n",
    "\n",
    "###############################\n",
    "###############################\n",
    "\n",
    "class LoadData(object):\n",
    "    \"\"\"Class to load data in uniform format\"\"\"\n",
    "    path = \"\" # Path to load from (usually folder)\n",
    "    output = True # Whether to print output\n",
    "    \n",
    "    def __init__(self, path=\"\"):\n",
    "        \"\"\"Can remember path if given\"\"\"\n",
    "        self.path=path\n",
    "    \n",
    "    def return_p(self, **kwargs):\n",
    "        \"\"\"Return array of Allele Frequencies [l]\"\"\"\n",
    "        raise NotImplementedError(\"Implement This in specific subclass.\")\n",
    "\n",
    "    def return_map(self, **kwargs):\n",
    "        \"\"\"Return genetic map [l] in Morgan\"\"\"\n",
    "        raise NotImplementedError(\"Implement This in specific subclass.\")\n",
    "        \n",
    "    def return_haplotypes_ll(self, **kwargs):\n",
    "        \"\"\"Return haplotype likelihoods [4,l,2]\"\"\"\n",
    "        raise NotImplementedError(\"Implement This in specific subclass.\")\n",
    "        \n",
    "    def load_all_data(self, **kwargs):\n",
    "        \"\"\"Key Method. \n",
    "        haplotype likelihoods [4,l,2]\n",
    "        derived allele frequencies [l]\n",
    "        map in Morgan [l]\"\"\"\n",
    "        htsl = self.return_haplotypes_ll()\n",
    "        p = self.return_p()\n",
    "        m = self.return_map()\n",
    "        self.check_valid_data(htsl, p, m)\n",
    "        return htsl, p, m\n",
    "    \n",
    "    def check_valid_data(self, htsl, p, m):\n",
    "        \"\"\"Check whether data in valid format\"\"\"\n",
    "        assert(np.shape(htsl)[1]==len(p))\n",
    "        assert(len(p)==len(m))\n",
    "        pass \n",
    "    \n",
    "    def rmap_to_gaps(self, r_map=[], cm=False, min_gap=1e-10, max_gap=0.05):\n",
    "        \"\"\"Return the recombination map gaps [in Morgan]\n",
    "        Input: Map Positions [l] units see cm below\n",
    "        Return: Rec. Distance Array [l]\n",
    "        cm: Whether input is in centimorgan or morgan\n",
    "        min_cap: Minimum Map Gap between Loci to cap\n",
    "        max_cap: Maximum Mapg Gap between Loci to cap\"\"\"\n",
    "        gaps = np.zeros(len(r_map))  # Make new array\n",
    "        gaps[1:] = r_map[1:] - r_map[:-1]  # Calculate Differences\n",
    "        assert(np.min(gaps) >= 0)\n",
    "        if cm == True:\n",
    "            gaps = gaps / 100     # Normalize to Morgan if map in cM\n",
    "         \n",
    "        ### Extend the minimum gap where needed\n",
    "        gaps = np.maximum(gaps, min_gap)\n",
    "        gaps = np.minimum(gaps, max_gap)\n",
    "\n",
    "        if self.output == True:\n",
    "            max_g = np.max(gaps)\n",
    "            print(f\"Minimum Genetic Map: {np.min(r_map):.4f} Morgan\")\n",
    "            print(f\"Maximum Genetic Map: {np.max(r_map):.4f} Morgan\")\n",
    "            print(f\"Gaps bigger than 0.1 cM: {np.sum(gaps > 0.001)}\")\n",
    "            print(f\"Maximum Gap: {max_g * 100:.4f} cM\")\n",
    "            print(f\"Upper Gap Cutoff: {max_gap * 100:.4f} cM\")\n",
    "        return gaps\n",
    "        \n",
    "    def set_params(self, **kwargs):\n",
    "        \"\"\"Set the Parameters.\n",
    "        Takes keyworded arguments\"\"\"\n",
    "        for key, value in kwargs.items():\n",
    "            setattr(self, key, value)\n",
    "            \n",
    "###############################\n",
    "### Class for Simulated date\n",
    "\n",
    "class LoadSimulated(LoadData):\n",
    "    \"\"\"Class to load simulated data saved in standard format.\"\"\"\n",
    "    r_gap = 1.0\n",
    "    r_path = \"\"\n",
    "    \n",
    "    def load_all_data(self, **kwargs):\n",
    "        ### Calculate Paths\n",
    "        haplo_path = os.path.join(self.path, \"haplo_ll.tsv\")\n",
    "        p_path = os.path.join(self.path, \"p.tsv\")\n",
    "        \n",
    "        ### Load the Data\n",
    "        #htsl = pickle.load(open(haplo_path, \"rb\")) \n",
    "        htsl = np.loadtxt(haplo_path, delimiter=\"\\t\", dtype=\"float\")\n",
    "        p = np.loadtxt(p_path, delimiter=\"\\t\", dtype=\"float\")\n",
    "        if len(self.r_path)==0:\n",
    "            m = np.ones(len(p), dtype=\"float\") * self.r_gap\n",
    "        \n",
    "        self.check_valid_data(htsl, p, m)\n",
    "        return htsl, p, m\n",
    "    \n",
    "###############################\n",
    "### Class for HDF5 data\n",
    "\n",
    "class LoadHDF5(LoadData):\n",
    "    \"\"\"Class to load simulated data saved in standard format.\"\"\"\n",
    "    path_h5=\"\"\n",
    "    iids = []\n",
    "    ch = 3   # Which chromosome to load\n",
    "    min_gap=1e-10 # Minimum Map Gap between two loci\n",
    "    max_gap=0.05  # Maximum Map Gap between two loci\n",
    "    min_error=1e-5 # Minimum Probability of genotyping erorr\n",
    "    \n",
    "    def return_map(self, f):\n",
    "        \"\"\"Return the recombination map\"\"\"\n",
    "        r_map = f[\"variants/MAP\"][:]\n",
    "        m = self.rmap_to_gaps(r_map=r_map, cm=False)\n",
    "        return m\n",
    "    \n",
    "    def return_p(self, f):\n",
    "        \"\"\"Return array of Allele Frequencies [l]\"\"\"\n",
    "        p = 0.5 * np.ones(len(f[\"variants/MAP\"]))\n",
    "        return p\n",
    "        \n",
    "    def get_individual_idx(self, f, iid, f_col=\"samples\"):\n",
    "        \"\"\"Return index of individual iid\"\"\"\n",
    "        samples = f[f_col][:]\n",
    "        idx = (samples == iids[0])\n",
    "        assert(np.sum(idx)==1) # Sanity Check\n",
    "        idx=np.where(idx)[0][0]\n",
    "        return idx  \n",
    "    \n",
    "    def get_haplo_prob(f, idx):\n",
    "        \"\"\"Get  haploid ancestral probability for indivual\"\"\"\n",
    "        h1 = f[\"calldata/GT\"][:,idx,:].T\n",
    "        m = np.max(f[\"calldata/GP\"][:,idx,:], axis=1)\n",
    "        m = np.minimum(m,  1 - self.min_error)\n",
    "        h1 = h1 * m + (1-h1) * (1 - m) # Probability of being ancestral\n",
    "        return h1\n",
    "        \n",
    "    def load_all_data(self, **kwargs):\n",
    "        \"\"\"Key Method. \n",
    "        haplotype likelihoods [4,l,2]\n",
    "        derived allele frequencies [l]\n",
    "        map in Morgan [l]\"\"\"\n",
    "        path_h5_ch = f\"{self.path}{self.ch}.h5\"\n",
    "        with h5py.File(path_h5_ch, \"r\") as f:\n",
    "            m = self.return_map(f)\n",
    "            p = self.return_p(f)\n",
    "            \n",
    "            idx1 = self.get_individual_idx(f, self.iids[0])\n",
    "            idx2 = self.get_individual_idx(f, self.iids[1])\n",
    "            h1 = get_haplo_prob(f, idx1)\n",
    "            h2 = get_haplo_prob(f, idx2)\n",
    "            htsl = np.concatenate((h1,h2), axis=0)\n",
    "        \n",
    "        self.check_valid_data(htsl, p, m)\n",
    "        return htsl, p, m\n",
    "    \n",
    "###############################    \n",
    "###############################\n",
    "### Factory Method\n",
    "    \n",
    "def load_loaddata(l_model=\"simulated\", path=\"\", **kwargs):\n",
    "    \"\"\"Factory method to return the right loading Model\"\"\"\n",
    "    if l_model == \"simulated\":\n",
    "        l_obj = LoadSimulated(path=path, **kwargs)\n",
    "    elif l_model == \"hdf5\":\n",
    "        l_obj = LoadHDF5(path=path, **kwargs)\n",
    "    else:\n",
    "        raise NotImplementedError(\"Loading Model not found!\")\n",
    "    return l_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test Loading the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load HDF5 Data for chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "ch=3\n",
    "iids = [\"SUC002\", \"SUC003\"]\n",
    "path_h5 = \"../../data/hdf5/1240k_v43/ch\"\n",
    "\n",
    "###\n",
    "path_h5_ch = f\"{path_h5}{ch}.h5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "l = load_loaddata(l_model=\"simulated\", path=\"../../output/simulated/undermodel/sim5/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 2.2326 Morgan\n",
      "Gaps bigger than 0.1 cM: 214\n",
      "Maximum Gap: 0.2348 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n"
     ]
    }
   ],
   "source": [
    "l = load_loaddata(l_model=\"hdf5\", path=\"../../data/hdf5/1240k_v43/ch\")\n",
    "l.set_params(iids=iids)\n",
    "hts_l, p, m = l.load_all_data()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
