{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to create the HDF5 File needed as Input for ancIBD\n",
    "\n",
    "This Vignette Notebook runs you through the steps for producing the input that ancIBD needs, a so called hdf5 file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Starting Point: A imputed and phased VCF\n",
    "The IBD software starts from an externally imputed and phased VCF. ancIBD is optimized to work with data that is output from `GLIMPSE` (see https://odelaneau.github.io/GLIMPSE/ for documentation and examples) and was imputed using the publicly available 1000G reference haplotype panel.\n",
    "\n",
    "### Transforming VCF to HDF5\n",
    "The notebook here showcases how to produce a HDF5 file from the Glimpse output VCF.\n",
    "\n",
    "### Requirement of input VCF:\n",
    "Importantly, the input VCF should have two fields:\n",
    "- GT: Diploid Genotype (the most likely imputed diploid genotype)\n",
    "- GP: Genotype probabilities\n",
    "\n",
    "These two fields are in the standard output of Glimpse. They get also transformed to the HDF5 file, and these data is key for a successful run of `ancIBD`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/n/groups/reich/hringbauer/git/hapBLOCK/notebook/vignette\n"
     ]
    }
   ],
   "source": [
    "### First do Imports\n",
    "import sys as sys\n",
    "import os\n",
    "import matplotlib.cm as cm\n",
    "import pandas as pd\n",
    "\n",
    "###\n",
    "# Edit the following path to your vignette folder:\n",
    "path = \"/n/groups/reich/hringbauer/git/hapBLOCK/notebook/vignette/\"\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "print(os.getcwd())\n",
    "###\n",
    "\n",
    "### The following Code sets the working directory to your ancIBD code \n",
    "### Please comment out and set path if you do not use the standard installation\n",
    "#sys.path.insert(0,\"/n/groups/reich/hringbauer/git/hapBLOCK/package/\")  # hack to get development package first in path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ancIBD.IO.prepare_h5 import vcf_to_1240K_hdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code to Transform HDF5 to VCF\n",
    "The funcion `vcf_to_1240k_hdf` runs the transformation from VCF and outputs a hdf5 for 1240k SNPs.\n",
    "\n",
    "### Input and Output:\n",
    "Generally, data is organized per chromosomes with one file for each chromosome.\n",
    "\n",
    "The example of `vcf_to_1240k_hdf` below transforms the full example vcf data in `./data/vcf.raw/`\n",
    "into hdf5 files suitable for `ancIBD` into  `./data/hdf5/`\n",
    "\n",
    "### Parameters\n",
    "One needs to set paths for the intermediate files:\n",
    "- path_vcf: Path of an intermediate VCF file - which is internally filtered to 1240k data \n",
    "- path_h5: Path of the output HDF5 files\n",
    "- marker_path: Path of the 1240k SNPs to use (a simple table provided with the example data)\n",
    "- map_path: Path of the map file to use (eigenstrat .snp provided with the example data, it has the map data included)\n",
    "- af_path (optional): Path of allele frequencies to merge into hdf5 file\n",
    "The data for the 1240k SNPs (`marker_path`), for the linkage map (`map_path`) and allele frequencies (`af_path`) are provided with the vignette, and only needs to be changed for custom SNP sets."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below you find the function call for a single chromosome. You can write a loop (see below), or also use an array job to run this function on a cluster. The latter will save a lot of runtime on big input data as it runs in parallel."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr22.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 15483 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 16420 Chr.22 1240K SNPs.\n",
      "Intersection 15408 out of 15483 HDF5 SNPs\n",
      "Interpolating 75 variants.\n",
      "Finished Chromosome 22.\n",
      "Adding map to HDF5...\n",
      "Intersection 15408 out of 15483 target HDF5 SNPs. 75 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr22.h5\n",
      "\n",
      "CPU times: user 7.62 s, sys: 527 ms, total: 8.14 s\n",
      "Wall time: 8.67 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "ch = 22\n",
    "\n",
    "base_path = f\"/n/groups/reich/hringbauer/git/hapBLOCK\"\n",
    "vcf_to_1240K_hdf(in_vcf_path = f\"./data/vcf.raw/example_hazelton_chr{ch}.vcf\",\n",
    "                 path_vcf = f\"./data/vcf.1240k/example_hazelton_chr{ch}.vcf\",\n",
    "                 path_h5 = f\"./data/hdf5/example_hazelton_chr{ch}.h5\",\n",
    "                 marker_path = f\"./data/filters/snps_bcftools_ch{ch}.csv\",\n",
    "                 map_path = f\"./data/v51.1_1240k.snp\", \n",
    "                 af_path = f\"./data/afs/v51.1_1240k_AF_ch{ch}.tsv\",\n",
    "                 col_sample_af = \"\", \n",
    "                 buffer_size=20000, chunk_width=8, chunk_length=20000,\n",
    "                 ch=ch)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loop Glimpse vcf -> hdf5 over all chromosomes\n",
    "This is the same as the run for a single chromsome above, but now looped over multiple chromsomes.\n",
    "Notice that the runtime can go into the minutes or even hours for larger datasets. Parallelization of this transformation (e.g. via a parralel array job on a cluster) can speed up that time if needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr1.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 88408 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 93166 Chr.1 1240K SNPs.\n",
      "Intersection 88115 out of 88408 HDF5 SNPs\n",
      "Interpolating 293 variants.\n",
      "Finished Chromosome 1.\n",
      "Adding map to HDF5...\n",
      "Intersection 88115 out of 88408 target HDF5 SNPs. 293 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr1.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr2.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 93875 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 98657 Chr.2 1240K SNPs.\n",
      "Intersection 93471 out of 93875 HDF5 SNPs\n",
      "Interpolating 404 variants.\n",
      "Finished Chromosome 2.\n",
      "Adding map to HDF5...\n",
      "Intersection 93471 out of 93875 target HDF5 SNPs. 404 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr2.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr3.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 77345 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 81416 Chr.3 1240K SNPs.\n",
      "Intersection 77013 out of 77345 HDF5 SNPs\n",
      "Interpolating 332 variants.\n",
      "Finished Chromosome 3.\n",
      "Adding map to HDF5...\n",
      "Intersection 77013 out of 77345 target HDF5 SNPs. 332 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr3.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr4.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 68518 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 71634 Chr.4 1240K SNPs.\n",
      "Intersection 68254 out of 68518 HDF5 SNPs\n",
      "Interpolating 264 variants.\n",
      "Finished Chromosome 4.\n",
      "Adding map to HDF5...\n",
      "Intersection 68254 out of 68518 target HDF5 SNPs. 264 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr4.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr5.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 69063 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 74004 Chr.5 1240K SNPs.\n",
      "Intersection 68899 out of 69063 HDF5 SNPs\n",
      "Interpolating 164 variants.\n",
      "Finished Chromosome 5.\n",
      "Adding map to HDF5...\n",
      "Intersection 68899 out of 69063 target HDF5 SNPs. 164 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr5.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr6.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 75347 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 78867 Chr.6 1240K SNPs.\n",
      "Intersection 75059 out of 75347 HDF5 SNPs\n",
      "Interpolating 288 variants.\n",
      "Finished Chromosome 6.\n",
      "Adding map to HDF5...\n",
      "Intersection 75059 out of 75347 target HDF5 SNPs. 288 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr6.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr7.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 59603 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 62595 Chr.7 1240K SNPs.\n",
      "Intersection 59324 out of 59603 HDF5 SNPs\n",
      "Interpolating 279 variants.\n",
      "Finished Chromosome 7.\n",
      "Adding map to HDF5...\n",
      "Intersection 59324 out of 59603 target HDF5 SNPs. 279 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr7.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr8.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 60828 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 63916 Chr.8 1240K SNPs.\n",
      "Intersection 60530 out of 60828 HDF5 SNPs\n",
      "Interpolating 298 variants.\n",
      "Finished Chromosome 8.\n",
      "Adding map to HDF5...\n",
      "Intersection 60530 out of 60828 target HDF5 SNPs. 298 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr8.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr9.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 50546 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 52765 Chr.9 1240K SNPs.\n",
      "Intersection 50307 out of 50546 HDF5 SNPs\n",
      "Interpolating 239 variants.\n",
      "Finished Chromosome 9.\n",
      "Adding map to HDF5...\n",
      "Intersection 50307 out of 50546 target HDF5 SNPs. 239 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr9.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr10.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 58610 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 61131 Chr.10 1240K SNPs.\n",
      "Intersection 58364 out of 58610 HDF5 SNPs\n",
      "Interpolating 246 variants.\n",
      "Finished Chromosome 10.\n",
      "Adding map to HDF5...\n",
      "Intersection 58364 out of 58610 target HDF5 SNPs. 246 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr10.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr11.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 54590 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 57163 Chr.11 1240K SNPs.\n",
      "Intersection 54365 out of 54590 HDF5 SNPs\n",
      "Interpolating 225 variants.\n",
      "Finished Chromosome 11.\n",
      "Adding map to HDF5...\n",
      "Intersection 54365 out of 54590 target HDF5 SNPs. 225 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr11.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr12.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 53737 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 56133 Chr.12 1240K SNPs.\n",
      "Intersection 53528 out of 53737 HDF5 SNPs\n",
      "Interpolating 209 variants.\n",
      "Finished Chromosome 12.\n",
      "Adding map to HDF5...\n",
      "Intersection 53528 out of 53737 target HDF5 SNPs. 209 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr12.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr13.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 38927 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 40441 Chr.13 1240K SNPs.\n",
      "Intersection 38774 out of 38927 HDF5 SNPs\n",
      "Interpolating 153 variants.\n",
      "Finished Chromosome 13.\n",
      "Adding map to HDF5...\n",
      "Intersection 38774 out of 38927 target HDF5 SNPs. 153 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr13.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr14.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 35885 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 37903 Chr.14 1240K SNPs.\n",
      "Intersection 35744 out of 35885 HDF5 SNPs\n",
      "Interpolating 141 variants.\n",
      "Finished Chromosome 14.\n",
      "Adding map to HDF5...\n",
      "Intersection 35744 out of 35885 target HDF5 SNPs. 141 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr14.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr15.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 34280 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 35991 Chr.15 1240K SNPs.\n",
      "Intersection 34159 out of 34280 HDF5 SNPs\n",
      "Interpolating 121 variants.\n",
      "Finished Chromosome 15.\n",
      "Adding map to HDF5...\n",
      "Intersection 34159 out of 34280 target HDF5 SNPs. 121 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr15.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr16.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 34335 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 36000 Chr.16 1240K SNPs.\n",
      "Intersection 34138 out of 34335 HDF5 SNPs\n",
      "Interpolating 198 variants.\n",
      "Finished Chromosome 16.\n",
      "Adding map to HDF5...\n",
      "Intersection 34138 out of 34335 target HDF5 SNPs. 197 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr16.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr17.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 28892 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 30733 Chr.17 1240K SNPs.\n",
      "Intersection 28794 out of 28892 HDF5 SNPs\n",
      "Interpolating 98 variants.\n",
      "Finished Chromosome 17.\n",
      "Adding map to HDF5...\n",
      "Intersection 28794 out of 28892 target HDF5 SNPs. 98 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr17.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr18.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 33846 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 35327 Chr.18 1240K SNPs.\n",
      "Intersection 33720 out of 33846 HDF5 SNPs\n",
      "Interpolating 126 variants.\n",
      "Finished Chromosome 18.\n",
      "Adding map to HDF5...\n",
      "Intersection 33720 out of 33846 target HDF5 SNPs. 126 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr18.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr19.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 18092 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 19273 Chr.19 1240K SNPs.\n",
      "Intersection 18018 out of 18092 HDF5 SNPs\n",
      "Interpolating 74 variants.\n",
      "Finished Chromosome 19.\n",
      "Adding map to HDF5...\n",
      "Intersection 18018 out of 18092 target HDF5 SNPs. 74 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr19.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr20.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 28940 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 30377 Chr.20 1240K SNPs.\n",
      "Intersection 28827 out of 28940 HDF5 SNPs\n",
      "Interpolating 113 variants.\n",
      "Finished Chromosome 20.\n",
      "Adding map to HDF5...\n",
      "Intersection 28827 out of 28940 target HDF5 SNPs. 113 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr20.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr21.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 15707 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 16727 Chr.21 1240K SNPs.\n",
      "Intersection 15640 out of 15707 HDF5 SNPs\n",
      "Interpolating 67 variants.\n",
      "Finished Chromosome 21.\n",
      "Adding map to HDF5...\n",
      "Intersection 15640 out of 15707 target HDF5 SNPs. 67 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr21.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Finished BCF tools filtering.\n",
      "Deleting previous HDF5 file at path_h5: ./data/hdf5/example_hazelton_chr22.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 15483 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 16420 Chr.22 1240K SNPs.\n",
      "Intersection 15408 out of 15483 HDF5 SNPs\n",
      "Interpolating 75 variants.\n",
      "Finished Chromosome 22.\n",
      "Adding map to HDF5...\n",
      "Intersection 15408 out of 15483 target HDF5 SNPs. 75 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: ./data/hdf5/example_hazelton_chr22.h5\n",
      "\n",
      "CPU times: user 2min 55s, sys: 9.35 s, total: 3min 4s\n",
      "Wall time: 5min 32s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "\n",
    "chs = range(1,23)\n",
    "\n",
    "for ch in chs:\n",
    "    base_path = f\"/n/groups/reich/hringbauer/git/hapBLOCK\"\n",
    "    vcf_to_1240K_hdf(in_vcf_path = f\"./data/vcf.raw/example_hazelton_chr{ch}.vcf.gz\",\n",
    "                     path_vcf = f\"./data/vcf.1240k/example_hazelton_chr{ch}.vcf\",\n",
    "                     path_h5 = f\"./data/hdf5/example_hazelton_chr{ch}.h5\",\n",
    "                     marker_path = f\"./data/filters/snps_bcftools_ch{ch}.csv\",\n",
    "                     map_path = f\"./data/v51.1_1240k.snp\", \n",
    "                     af_path = f\"./data/afs/v51.1_1240k_AF_ch{ch}.tsv\",\n",
    "                     col_sample_af = \"\", \n",
    "                     buffer_size=20000, chunk_width=8, chunk_length=20000,\n",
    "                     ch=ch)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Next Steps\n",
    "This is it, congratulations! Now you have the output data needed for `ancIBD`. Now you can continue with the vignette notebook in `./run_ancIBD.ipynb`."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
