{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create test data to check implementation of HMM\n",
    "Creates data under the model!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "compute-a-17-93.o2.rc.hms.harvard.edu\n",
      "HSM O2 Computational partition detected.\n",
      "/n/groups/reich/hringbauer/git/hapBLOCK\n",
      "CPU Count: 32\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import socket as socket\n",
    "import os as os\n",
    "import sys as sys\n",
    "import multiprocessing as mp\n",
    "import h5py\n",
    "import allel\n",
    "import itertools as it\n",
    "import pickle\n",
    "\n",
    "socket_name = socket.gethostname()\n",
    "print(socket_name)\n",
    "\n",
    "if socket_name.startswith(\"compute-\"):\n",
    "    print(\"HSM O2 Computational partition detected.\")\n",
    "    path = \"/n/groups/reich/hringbauer/git/hapBLOCK/\"  # The Path on Harvard Cluster\n",
    "else: \n",
    "    raise RuntimeWarning(\"Not compatible machine. Check!!\")\n",
    "\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "\n",
    "print(os.getcwd())\n",
    "print(f\"CPU Count: {mp.cpu_count()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Array of Means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_normal_all_freqs(l=5000, mean=0.5, std=0.1):\n",
    "    \"\"\"Draw Normally distributed allele frequencies\"\"\"\n",
    "    p = np.random.normal(size=l, loc=mean, scale=std)\n",
    "    return p\n",
    "\n",
    "def draw_haplotype(p):\n",
    "    \"\"\"Draw binomal haplotype from allele frequency p.\n",
    "    For each locus draw one genotype\n",
    "    Assumes Hardy Weinbgerg.\"\"\"\n",
    "    l = len(p)\n",
    "    s = np.random.random(size=l)<p\n",
    "    return s\n",
    "\n",
    "def create_hap_ll(h):\n",
    "    \"\"\"Create vector of haplotype likelihood.\n",
    "    Without error. \n",
    "    h: Haplotype vector [l,2] consisting of 0/1 post. prob\"\"\"\n",
    "    l = np.zeros((len(h),2), dtype=\"bool\")\n",
    "    idx_der = (h==1)\n",
    "    l[idx_der, 1] = 1\n",
    "    l[~idx_der, 0] = 1\n",
    "    return l\n",
    "\n",
    "def add_haplo_rnd_error(h, eps=0.01):\n",
    "    \"\"\"Modify array of haplotyp post [l,2].\n",
    "    Add random error and return array\n",
    "    [l,2] of new haplotype posterior\"\"\"\n",
    "    h = h * (1 - eps) + np.roll(h, shift=1, axis=1) * eps\n",
    "    return h\n",
    "    \n",
    "def copy_in_block(h_s, h_t, loc=[10,20]):\n",
    "    \"\"\"Copy in Haplotype block by copying segment.\n",
    "    h_s: The source haplotype\n",
    "    h_t: The target haplotype\n",
    "    loc: Location of haplotype to copy over\"\"\"\n",
    "    h_t[loc[0]:loc[1]]=h_s[loc[0]:loc[1]]\n",
    "    return h_t\n",
    "\n",
    "def create_ibd_haplos(l=10000, loc=[2000,4000], mean=0.5, std=0.1, eps=0):\n",
    "    \"\"\"Create haplotype likelihoods with IBD copying.\n",
    "    Draw allele frequencies, create HW haplotypes, copy in haplotype\n",
    "    and create likelihoods.\n",
    "    Return p [l] and haplotypelikelihoods [4,l,2]\"\"\"\n",
    "    p = draw_normal_all_freqs(l=l, mean=mean, std=std)\n",
    "    hts = [draw_haplotype(p) for _ in range(4)]\n",
    "    hts[0] = copy_in_block(h_s=hts[2], h_t=hts[0], loc=loc)\n",
    "    htsl = np.array([create_hap_ll(h) for h in hts])\n",
    "    htsl = np.array([add_haplo_rnd_error(h, eps=eps) for h in htsl])\n",
    "    return p, htsl\n",
    "\n",
    "def save_haplo_ll(hts, p=[], folder=\"./output/simulated/undermodel/\",\n",
    "                  delimiter='\\t'):\n",
    "    \"\"\"Save Haplotype Likelihoods in standardized format.\"\"\"\n",
    "    if not os.path.exists(folder):\n",
    "        os.makedirs(folder)\n",
    "    savepath = os.path.join(folder,\"haplo_ll.tsv\")\n",
    "    pickle.dump(hts, open(savepath, \"wb\") )\n",
    "    #np.savetxt(savepath, hts, delimiter=delimiter)\n",
    "    print(f\"Saved {np.shape(hts)[1]} loci likelihoods to {savepath}\")\n",
    "    \n",
    "    if len(p)>0:\n",
    "        savepath = os.path.join(folder,\"p.tsv\")\n",
    "        np.savetxt(savepath, p, delimiter=delimiter)\n",
    "        print(f\"Saved {len(p)} allele frequencies to {savepath}\")\n",
    "\n",
    "def shuffle_haplos(h1, h2, scale=20):\n",
    "    \"\"\"Shuffle hapolotypes with random waiting times \n",
    "    (mean scale, exponentially distributed)\"\"\"\n",
    "    hn1, hn2 = np.copy(h1), np.copy(h2)\n",
    "    l=len(h1)\n",
    "    w = np.random.exponential(scale=scale,size=l)\n",
    "    w = np.round(w).astype(\"int\")\n",
    "    w = np.cumsum(w)\n",
    "    s=0\n",
    "    for i in range(len(w)):\n",
    "        a,b=w[i],w[i+1]\n",
    "        #print(f\"shuffling from {a}-{b}\")\n",
    "        if b>l:\n",
    "            break\n",
    "        if (i%2)==1: # Only every other is shuffled\n",
    "            hn1[a:b],hn2[a:b] = h2[a:b],h1[a:b]\n",
    "        s+=1\n",
    "    print(f\"Total shuffles: {s}\")\n",
    "    return hn1,hn2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test single Haplotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of derived variants: 4910 / 10000\n",
      "Number of derived variants haplotype ll: 4910 / 10000\n"
     ]
    }
   ],
   "source": [
    "p = draw_normal_all_freqs(l=10000, mean=0.5, std=0.1)\n",
    "assert((np.min(p)>0) and (np.max(p)<1)) # Sanity Check\n",
    "\n",
    "### Draw random Haplotype\n",
    "h0 = draw_haplotype(p)\n",
    "print(f\"Number of derived variants: {np.sum(h0)} / {len(h0)}\")\n",
    "l0 = create_hap_ll(h0)\n",
    "print(f\"Number of derived variants haplotype ll: {np.sum(l0[:,1])} / {len(l0)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create IBD sharing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6.97 ms, sys: 1.17 ms, total: 8.13 ms\n",
      "Wall time: 6.51 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "p, htsl = create_ibd_haplos(l=10000, loc=[2000,4000], mean=0.5, std=0.1, eps=0.05)\n",
    "np.sum(htsl,axis=2).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total shuffles: 990\n",
      "Total shuffles: 984\n"
     ]
    }
   ],
   "source": [
    "### Shuffle Haplotypes if needed, to create phase switches\n",
    "htsl[0],htsl[1] = shuffle_haplos(htsl[0],htsl[1], scale=10)\n",
    "htsl[2],htsl[3] = shuffle_haplos(htsl[2],htsl[3], scale=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved 10000 loci likelihoods to ./output/simulated/undermodel/sim5/haplo_ll.tsv\n",
      "Saved 10000 allele frequencies to ./output/simulated/undermodel/sim5/p.tsv\n"
     ]
    }
   ],
   "source": [
    "save_haplo_ll(hts=htsl, p=p, folder=\"./output/simulated/undermodel/sim5/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Area 51"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test load the saved the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#htsl1 = pickle.load(\"./output/simulated/undermodel/sim1/haplo_ll.tsv\", delimiter=\"\\t\", dtype=\"float\")\n",
    "htsl1 = pickle.load(open(\"./output/simulated/undermodel/sim1/haplo_ll.tsv\", \"rb\")) \n",
    "#htsl1 = np.loadtxt(\"./output/simulated/undermodel/sim1/haplo_ll.tsv\", delimiter=\"\\t\", dtype=\"float\")\n",
    "p1 = np.loadtxt(\"./output/simulated/undermodel/sim1/p.tsv\", delimiter=\"\\t\", dtype=\"float\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean, std = 0.5, 0.1\n",
    "p = draw_normal_all_freqs(l=10000, mean=mean, std=std)\n",
    "hts = [draw_haplotype(p) for _ in range(4)]\n",
    "hts[0] = copy_in_block(h_s=hts[2], h_t=hts[0], loc=[2000,4000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "htsl = np.array([create_hap_ll(h) for h in hts])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "t = htsl[0,:,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([3.3000e+01, 1.3200e+02, 1.3800e+02, ..., 3.9310e+04, 3.9342e+04,\n",
       "       3.9344e+04])"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "h0, h1 = htsl[0], htsl[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10000, 2)"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(h0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total shuffles: 19\n"
     ]
    }
   ],
   "source": [
    "h1 = np.zeros(1000)\n",
    "h2 = np.ones(1000)\n",
    "\n",
    "h1,h2 = shuffle_haplos(h1,h2,scale=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7feeb7545510>]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAARtklEQVR4nO3da4yc113H8e/Pu3Ucc+klXqoSJ9lUmMumApKsQqoiiLApaUDJCy6KBaJAhKW4hQIVKBGohaAIlSJuIkADFCRsEkJBYIWABSG8QTRko0LIpWndNBeHtllKKRJVaV3+vJhnyXizl9ndZ3d2j78f6ZHnOc+Z55wzZ+a3s8/Z8aSqkCTtfLvG3QFJUj8MdElqhIEuSY0w0CWpEQa6JDViclwN79u3r6anp8fVvCTtSA8//PC/V9XUUsfGFujT09PMzc2Nq3lJ2pGSPLPcMS+5SFIjDHRJaoSBLkmNMNAlqREGuiQ1YtVAT/LeJC8keXSZ40ny60lOJXkkyRX9d7Nz/Dicdx4k22s7enTThnyWQ4dG689ll/U7vpe9bPDY9227zudK26FDZ4/h6NHxtH/8OExOru2+u3at7T59Pq9Hfe6eS9v55/f/uqqqFTfgm4ArgEeXOX4d8FdAgKuBB1c7Z1Vx5ZVX1pocO1YF23e7+ea1jWetDh4c7/iSwRz0ZbvP50rbwYODMdx883jan5nZurb6eF6P+7m7nbddu9b8ugLmqpbO1QyOryzJNHBvVb1uiWPvAf6+qu7q9p8Erqmqj610ztnZ2VrT36FPT8Mzz4xef6tNTMCZM5t3/mTzzj2qSy6Bp5/u51zbfT5XUzV4t/uFL4y7J5urj+f1dnjubmdrfF0lebiqZpc61sc19AuB54b2T3dlS3XkSJK5JHPz8/Nra+XZZ9fdwS3R+gsb+p2D7T6fozgX5vxcGOO49fha2NJF0aq6s6pmq2p2amrJT64u7+KLN6dTfZmYGHcPNl+fc7Dd53MU58KcnwtjHLceXwt9BPrzwEVD+/u7sn7dfnvvp+zVkSObe/6DBzf3/KtJ+p2D7T6fK1mYi82e8+XMzGxdW32McdzP3e1s165+XwvLXVwf3oBpll8U/XbOXhT9p1HOueZF0arB4sHu3eNfyNiMhaNRjLq41Pei2eRkvwuiC7brfK60LSyILtjqhdGF9o8dq5qYWNt9k7Xdp8/ntQujL9327FnX64qNLIomuQu4BtgHfAJ4J/Cy7ofBbycJ8BvAtcBngB+sqlVXO9e8KCpJWnFRdNX/bbGqDq9yvIC3rLNvkqSe+ElRSWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaMVKgJ7k2yZNJTiW5ZYnjFyd5IMkHkjyS5Lr+uypJWsmqgZ5kArgDeBMwAxxOMrOo2s8A91TV5cCNwG/23VFJ0spGeYd+FXCqqp6qqs8BdwM3LKpTwJd2t18O/Ft/XZQkjWKUQL8QeG5o/3RXNuxnge9Lchq4D/iRpU6U5EiSuSRz8/Pz6+iuJGk5fS2KHgb+oKr2A9cBf5jkJeeuqjuraraqZqempnpqWpIEowX688BFQ/v7u7JhNwH3AFTVPwJ7gH19dFCSNJpRAv0h4ECSS5PsZrDoeWJRnWeBgwBJvoZBoHtNRZK20KqBXlVngLcCJ4EnGPw1y2NJbktyfVft7cAPJ/kX4C7gB6qqNqvTkqSXmhylUlXdx2Cxc7jsHUO3Hwfe0G/XJElr4SdFJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiNGCvQk1yZ5MsmpJLcsU+d7kjye5LEkf9RvNyVJq5lcrUKSCeAO4FuB08BDSU5U1eNDdQ4AtwJvqKpPJfmyzeqwJGlpo7xDvwo4VVVPVdXngLuBGxbV+WHgjqr6FEBVvdBvNyVJqxkl0C8EnhvaP92VDftK4CuT/EOS9ye5dqkTJTmSZC7J3Pz8/Pp6LElaUl+LopPAAeAa4DDwO0lesbhSVd1ZVbNVNTs1NdVT05IkGC3QnwcuGtrf35UNOw2cqKrPV9VHgQ8xCHhJ0hYZJdAfAg4kuTTJbuBG4MSiOn/O4N05SfYxuATzVI/9lCStYtVAr6ozwFuBk8ATwD1V9ViS25Jc31U7CXwyyePAA8BPVtUnN6vTkqSXSlWNpeHZ2dmam5sbS9uStFMlebiqZpc65idFJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqxEiBnuTaJE8mOZXklhXqfWeSSjLbXxclSaNYNdCTTAB3AG8CZoDDSWaWqPclwNuAB/vupCRpdaO8Q78KOFVVT1XV54C7gRuWqPfzwLuAz/bYP0nSiEYJ9AuB54b2T3dl/y/JFcBFVfWXK50oyZEkc0nm5ufn19xZSdLyNrwommQX8MvA21erW1V3VtVsVc1OTU1ttGlJ0pBRAv154KKh/f1d2YIvAV4H/H2Sp4GrgRMujErS1hol0B8CDiS5NMlu4EbgxMLBqvp0Ve2rqumqmgbeD1xfVXOb0mNJ0pJWDfSqOgO8FTgJPAHcU1WPJbktyfWb3UFJ0mgmR6lUVfcB9y0qe8cyda/ZeLckSWvlJ0UlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSI0YK9CTXJnkyyakktyxx/CeSPJ7kkST3J7mk/65KklayaqAnmQDuAN4EzACHk8wsqvYBYLaqvhZ4H/CLfXdUkrSyUd6hXwWcqqqnqupzwN3ADcMVquqBqvpMt/t+YH+/3ZQkrWaUQL8QeG5o/3RXtpybgL9a6kCSI0nmkszNz8+P3ktJ0qp6XRRN8n3ALPDupY5X1Z1VNVtVs1NTU302LUnnvMkR6jwPXDS0v78rO0uSQ8BPA99cVf/TT/ckSaMa5R36Q8CBJJcm2Q3cCJwYrpDkcuA9wPVV9UL/3ZQkrWbVQK+qM8BbgZPAE8A9VfVYktuSXN9VezfwxcCfJPnnJCeWOZ0kaZOMcsmFqroPuG9R2TuGbh/quV+SpDXyk6KS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDVipEBPcm2SJ5OcSnLLEsfPS/LH3fEHk0z33dFt5ehRSMa7HTr0Yn+OH4fzzlu63vnnD/q7b9+LZfv2De6z3cZ49Gi/83T8OExPw65dg39XG/NyLrtsvGNYbm5HaXeleVx8n1Hb2uh2/vnrnwutrKpW3IAJ4CPAa4HdwL8AM4vqHAV+u7t9I/DHq533yiuvrB3p5purYHtsBw9WHTu2vvvu3j2473Yb48039zNPx45V7d179rn37l1+zMuZmRnvGDbS7ijzuHCf9T6P1rvt2rX2uVBVVQFzVUvnagbHl5fk9cDPVtW3dfu3dj8IfmGozsmuzj8mmQQ+DkzVCiefnZ2tubm5df0QGqvJSfjCF8bdixddcgk888z67/v00y8tH+cYJybgzJmNn2d6eunHZbkxLydZe9ubPYZR2x1lHhfus9a2+rDWuRAASR6uqtmljo1yyeVC4Lmh/dNd2ZJ1quoM8GnggiU6ciTJXJK5+fn5Ufq+/WynMAd49tn+7zvOMfbV9nJj28jjNarNHsOo7Y7Sj4U6W/G4LDaONhu3pYuiVXVnVc1W1ezU1NRWNt2fiYlx9+BsF1/c/33HOca+2l5ubBt5vEa12WMYtd1R+rFQZysel8XG0WbjRgn054GLhvb3d2VL1ukuubwc+GQfHdx2jhwZdw9edPAg3H77+u67e/fy9x3nGPtq+/bbYe/es8v27l374zUzs/a2+xzDRtodpR8Lddb7PFqvXbu2vs1zwXIX1xc2YBJ4CriUFxdFL1tU5y2cvSh6z2rn3bGLolXbY2H04MEX+3Ps2GCRc6l6e/YM+nvBBS+WXXDB6gtS4xhjX4uJw4/LJZdUJYN/17sIt5aF0c0Yw3JzO0q7K83j4vuM2tZGtz17XBDdADayKAqQ5DrgVxn8xct7q+r2JLd1Jz6RZA/wh8DlwH8AN1bVUyudc8cuikrSGK20KDo5ygmq6j7gvkVl7xi6/VnguzfSSUnSxvhJUUlqhIEuSY0w0CWpEQa6JDVipL9y2ZSGk3lgvZ813gf8e4/d2Qkc87nBMZ8bNjLmS6pqyU9mji3QNyLJ3HJ/ttMqx3xucMznhs0as5dcJKkRBrokNWKnBvqd4+7AGDjmc4NjPjdsyph35DV0SdJL7dR36JKkRQx0SWrEjgv01b6weqdKclGSB5I8nuSxJG/ryl+V5G+SfLj795VdeZL8evc4PJLkivGOYH2STCT5QJJ7u/1Luy8aP9V98fjurryJLyJP8ook70vywSRPJHn9OTDHP949px9NcleSPS3Oc5L3JnkhyaNDZWue2yRv7up/OMmb19KHHRXoSSaAO4A3ATPA4STr+AaCbekM8PaqmgGuBt7Sje0W4P6qOgDc3+3D4DE40G1HgN/a+i734m3AE0P77wJ+paq+AvgUcFNXfhPwqa78V7p6O9GvAX9dVV8NfB2DsTc7x0kuBH4UmK2q1zH4L7hvpM15/gPg2kVla5rbJK8C3gl8A3AV8M6FHwIjWe4/St+OG/B64OTQ/q3ArePu1yaN9S+AbwWeBF7Tlb0GeLK7/R7g8FD9/6+3UzYG3351P/AtwL1AGHx6bnLxfAMngdd3tye7ehn3GNY43pcDH13c78bneOH7hl/Vzdu9wLe1Os/ANPDoeucWOAy8Z6j8rHqrbTvqHTqjfWH1jtf9mnk58CDw6qr6WHfo48Cru9stPBa/CvwU8L/d/gXAf9bgi8bh7DGN9EXk29ylwDzw+91lpt9N8kU0PMdV9TzwS8CzwMcYzNvDtD3Pw9Y6txua850W6M1L8sXAnwI/VlX/NXysBj+ym/g70yTfAbxQVQ+Puy9baBK4Avitqroc+G9e/BUcaGuOAbrLBTcw+GH25cAX8dLLEueErZjbnRboo3xh9Y6V5GUMwvx4Vf1ZV/yJJK/pjr8GeKEr3+mPxRuA65M8DdzN4LLLrwGv6L5oHM4eUwtfRH4aOF1VD3b772MQ8K3OMcAh4KNVNV9Vnwf+jMHctzzPw9Y6txua850W6A8BB7oV8t0MFldOjLlPvUgS4PeAJ6rql4cOnQAWVrrfzODa+kL593er5VcDnx761W7bq6pbq2p/VU0zmMe/q6rvBR4Avqurtni8C4/Dd3X1d9Q72ar6OPBckq/qig4Cj9PoHHeeBa5Osrd7ji+Mudl5XmStc3sSeGOSV3a/3byxKxvNuBcR1rHocB3wIeAjwE+Puz89jusbGfw69gjwz912HYPrh/cDHwb+FnhVVz8M/uLnI8C/MvgrgrGPY51jvwa4t7v9WuCfgFPAnwDndeV7uv1T3fHXjrvf6xzr1wNz3Tz/OfDK1ucY+Dngg8CjDL5M/rwW5xm4i8E6wecZ/DZ203rmFvihbvyngB9cSx/86L8kNWKnXXKRJC3DQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmN+D+4bp8d5F2U2AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure()\n",
    "ax=plt.gca()\n",
    "\n",
    "ax.plot(np.arange(len(h1)), h2, \"ro\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
