{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick Start for ancIBD\n",
    "\n",
    "This notebook is a quick start guide for running ancIBD. It uses wrapper scripts for various functions introduced in section [preparing input](create_hdf5_from_vcf.ipynb) and [calling IBD with ancIBD](run_ancIBD.ipynb). Writing your own wrapper script for these functions provides more flexibility, while using the command line interface to be introduced in this quick starting guide is easier. We have created two command-line interfaces (`ancIBD-run` and `ancIBD-summary`)for running ancIBD quickly on your imputed data. The test data used to run these tutorials can be downloaded from https://www.dropbox.com/sh/q18yyrffbdj1yv1/AAC1apifYB_oKB8SNrmQQ-26a?dl=0. It contains imputed vcf of a subset of samples from early Neolithic Britain that belong to an extended pedigree ([Fowler et al.](https://www.nature.com/articles/s41586-021-04241-4)). "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### calling IBD\n",
    "\n",
    "In addition to the imputed vcf files, you need additionally three files, all of which are provided in the same dropbox link as indicated above.\n",
    "\n",
    "* marker_path: Path of the 1240k SNPs to use (you can find those in `./filters/snps_bcftools_ch*.csv` from the download link)\n",
    "* map_path: Path of the map file to use (eigenstrat .snp file, you can find it in `./afs/v51.1_1240k.snp` from the download link)\n",
    "* af_path (optional): Path of allele frequencies to merge into hdf5 file (you can find it in `./afs/v51.1_1240k_AF_ch*.tsv` from the download link. If not provided, allele frequencies calculated from samples themselves will be used)\n",
    "\n",
    "We now run ancIBD on ch20 as an example. To run the following command, change the path to the above three files according to your own environment if needed. The file path in the following tutorial has assumed that the folder downloaded from dropbox link is in the same directory as this jupyter notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modify file paths according to your own environment if needed\n",
    "ch = 20\n",
    "marker_path = f'./data/filters/snps_bcftools_ch{ch}.csv'\n",
    "map_path = './data/afs/v51.1_1240k.snp'\n",
    "af_path = f'./data/afs/v51.1_1240k_AF_ch{ch}.tsv'\n",
    "vcf_path = f'./data/vcf.raw/example_hazelton_chr{ch}.vcf.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch20.1240k.vcf -T ./data/filters/snps_bcftools_ch20.csv -M2 -v snps ./data/vcf.raw/example_hazelton_chr20.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 28940 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 30377 Chr.20 1240K SNPs.\n",
      "Intersection 28827 out of 28940 HDF5 SNPs\n",
      "Interpolating 113 variants.\n",
      "Finished Chromosome 20.\n",
      "Adding map to HDF5...\n",
      "Intersection 28827 out of 28940 target HDF5 SNPs. 113 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch20.h5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!ancIBD-run --vcf $vcf_path --ch $ch --out test --marker_path $marker_path --map_path $map_path --af_path $af_path --prefix example_hazelton"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you already have the appropriate hdf5 file for your samples, you can also supply the command line with the hdf5 file directly. But please make sure that the hdf5 file has suffix \"ch{chromosome number}.h5\" (e.g, \"test.ch20.h5\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ancIBD-run --h5 ./test/example_hazelton.ch20.h5 --ch $ch --out test --marker_path $marker_path --map_path $map_path --af_path $af_path --prefix example_hazelton"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "now we can do the same for the all the 22 autosomes. This takes about 6min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch1.1240k.vcf -T data/filters/snps_bcftools_ch1.csv -M2 -v snps data/vcf.raw/example_hazelton_chr1.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 88408 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 93166 Chr.1 1240K SNPs.\n",
      "Intersection 88115 out of 88408 HDF5 SNPs\n",
      "Interpolating 293 variants.\n",
      "Finished Chromosome 1.\n",
      "Adding map to HDF5...\n",
      "Intersection 88115 out of 88408 target HDF5 SNPs. 293 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch1.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch2.1240k.vcf -T data/filters/snps_bcftools_ch2.csv -M2 -v snps data/vcf.raw/example_hazelton_chr2.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 93875 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 98657 Chr.2 1240K SNPs.\n",
      "Intersection 93471 out of 93875 HDF5 SNPs\n",
      "Interpolating 404 variants.\n",
      "Finished Chromosome 2.\n",
      "Adding map to HDF5...\n",
      "Intersection 93471 out of 93875 target HDF5 SNPs. 404 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch2.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch3.1240k.vcf -T data/filters/snps_bcftools_ch3.csv -M2 -v snps data/vcf.raw/example_hazelton_chr3.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 77345 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 81416 Chr.3 1240K SNPs.\n",
      "Intersection 77013 out of 77345 HDF5 SNPs\n",
      "Interpolating 332 variants.\n",
      "Finished Chromosome 3.\n",
      "Adding map to HDF5...\n",
      "Intersection 77013 out of 77345 target HDF5 SNPs. 332 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch3.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch4.1240k.vcf -T data/filters/snps_bcftools_ch4.csv -M2 -v snps data/vcf.raw/example_hazelton_chr4.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 68518 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 71634 Chr.4 1240K SNPs.\n",
      "Intersection 68254 out of 68518 HDF5 SNPs\n",
      "Interpolating 264 variants.\n",
      "Finished Chromosome 4.\n",
      "Adding map to HDF5...\n",
      "Intersection 68254 out of 68518 target HDF5 SNPs. 264 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch4.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch5.1240k.vcf -T data/filters/snps_bcftools_ch5.csv -M2 -v snps data/vcf.raw/example_hazelton_chr5.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 69063 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 74004 Chr.5 1240K SNPs.\n",
      "Intersection 68899 out of 69063 HDF5 SNPs\n",
      "Interpolating 164 variants.\n",
      "Finished Chromosome 5.\n",
      "Adding map to HDF5...\n",
      "Intersection 68899 out of 69063 target HDF5 SNPs. 164 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch5.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch6.1240k.vcf -T data/filters/snps_bcftools_ch6.csv -M2 -v snps data/vcf.raw/example_hazelton_chr6.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 75347 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 78867 Chr.6 1240K SNPs.\n",
      "Intersection 75059 out of 75347 HDF5 SNPs\n",
      "Interpolating 288 variants.\n",
      "Finished Chromosome 6.\n",
      "Adding map to HDF5...\n",
      "Intersection 75059 out of 75347 target HDF5 SNPs. 288 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch6.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch7.1240k.vcf -T data/filters/snps_bcftools_ch7.csv -M2 -v snps data/vcf.raw/example_hazelton_chr7.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 59603 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 62595 Chr.7 1240K SNPs.\n",
      "Intersection 59324 out of 59603 HDF5 SNPs\n",
      "Interpolating 279 variants.\n",
      "Finished Chromosome 7.\n",
      "Adding map to HDF5...\n",
      "Intersection 59324 out of 59603 target HDF5 SNPs. 279 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch7.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch8.1240k.vcf -T data/filters/snps_bcftools_ch8.csv -M2 -v snps data/vcf.raw/example_hazelton_chr8.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 60828 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 63916 Chr.8 1240K SNPs.\n",
      "Intersection 60530 out of 60828 HDF5 SNPs\n",
      "Interpolating 298 variants.\n",
      "Finished Chromosome 8.\n",
      "Adding map to HDF5...\n",
      "Intersection 60530 out of 60828 target HDF5 SNPs. 298 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch8.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch9.1240k.vcf -T data/filters/snps_bcftools_ch9.csv -M2 -v snps data/vcf.raw/example_hazelton_chr9.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 50546 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 52765 Chr.9 1240K SNPs.\n",
      "Intersection 50307 out of 50546 HDF5 SNPs\n",
      "Interpolating 239 variants.\n",
      "Finished Chromosome 9.\n",
      "Adding map to HDF5...\n",
      "Intersection 50307 out of 50546 target HDF5 SNPs. 239 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch9.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch10.1240k.vcf -T data/filters/snps_bcftools_ch10.csv -M2 -v snps data/vcf.raw/example_hazelton_chr10.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 58610 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 61131 Chr.10 1240K SNPs.\n",
      "Intersection 58364 out of 58610 HDF5 SNPs\n",
      "Interpolating 246 variants.\n",
      "Finished Chromosome 10.\n",
      "Adding map to HDF5...\n",
      "Intersection 58364 out of 58610 target HDF5 SNPs. 246 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch10.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch11.1240k.vcf -T data/filters/snps_bcftools_ch11.csv -M2 -v snps data/vcf.raw/example_hazelton_chr11.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 54590 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 57163 Chr.11 1240K SNPs.\n",
      "Intersection 54365 out of 54590 HDF5 SNPs\n",
      "Interpolating 225 variants.\n",
      "Finished Chromosome 11.\n",
      "Adding map to HDF5...\n",
      "Intersection 54365 out of 54590 target HDF5 SNPs. 225 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch11.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch12.1240k.vcf -T data/filters/snps_bcftools_ch12.csv -M2 -v snps data/vcf.raw/example_hazelton_chr12.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 53737 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 56133 Chr.12 1240K SNPs.\n",
      "Intersection 53528 out of 53737 HDF5 SNPs\n",
      "Interpolating 209 variants.\n",
      "Finished Chromosome 12.\n",
      "Adding map to HDF5...\n",
      "Intersection 53528 out of 53737 target HDF5 SNPs. 209 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch12.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch13.1240k.vcf -T data/filters/snps_bcftools_ch13.csv -M2 -v snps data/vcf.raw/example_hazelton_chr13.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 38927 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 40441 Chr.13 1240K SNPs.\n",
      "Intersection 38774 out of 38927 HDF5 SNPs\n",
      "Interpolating 153 variants.\n",
      "Finished Chromosome 13.\n",
      "Adding map to HDF5...\n",
      "Intersection 38774 out of 38927 target HDF5 SNPs. 153 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch13.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch14.1240k.vcf -T data/filters/snps_bcftools_ch14.csv -M2 -v snps data/vcf.raw/example_hazelton_chr14.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 35885 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 37903 Chr.14 1240K SNPs.\n",
      "Intersection 35744 out of 35885 HDF5 SNPs\n",
      "Interpolating 141 variants.\n",
      "Finished Chromosome 14.\n",
      "Adding map to HDF5...\n",
      "Intersection 35744 out of 35885 target HDF5 SNPs. 141 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch14.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch15.1240k.vcf -T data/filters/snps_bcftools_ch15.csv -M2 -v snps data/vcf.raw/example_hazelton_chr15.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 34280 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 35991 Chr.15 1240K SNPs.\n",
      "Intersection 34159 out of 34280 HDF5 SNPs\n",
      "Interpolating 121 variants.\n",
      "Finished Chromosome 15.\n",
      "Adding map to HDF5...\n",
      "Intersection 34159 out of 34280 target HDF5 SNPs. 121 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch15.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch16.1240k.vcf -T data/filters/snps_bcftools_ch16.csv -M2 -v snps data/vcf.raw/example_hazelton_chr16.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 34335 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 36000 Chr.16 1240K SNPs.\n",
      "Intersection 34138 out of 34335 HDF5 SNPs\n",
      "Interpolating 198 variants.\n",
      "Finished Chromosome 16.\n",
      "Adding map to HDF5...\n",
      "Intersection 34138 out of 34335 target HDF5 SNPs. 197 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch16.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch17.1240k.vcf -T data/filters/snps_bcftools_ch17.csv -M2 -v snps data/vcf.raw/example_hazelton_chr17.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 28892 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 30733 Chr.17 1240K SNPs.\n",
      "Intersection 28794 out of 28892 HDF5 SNPs\n",
      "Interpolating 98 variants.\n",
      "Finished Chromosome 17.\n",
      "Adding map to HDF5...\n",
      "Intersection 28794 out of 28892 target HDF5 SNPs. 98 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch17.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch18.1240k.vcf -T data/filters/snps_bcftools_ch18.csv -M2 -v snps data/vcf.raw/example_hazelton_chr18.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 33846 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 35327 Chr.18 1240K SNPs.\n",
      "Intersection 33720 out of 33846 HDF5 SNPs\n",
      "Interpolating 126 variants.\n",
      "Finished Chromosome 18.\n",
      "Adding map to HDF5...\n",
      "Intersection 33720 out of 33846 target HDF5 SNPs. 126 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch18.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch19.1240k.vcf -T data/filters/snps_bcftools_ch19.csv -M2 -v snps data/vcf.raw/example_hazelton_chr19.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 18092 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 19273 Chr.19 1240K SNPs.\n",
      "Intersection 18018 out of 18092 HDF5 SNPs\n",
      "Interpolating 74 variants.\n",
      "Finished Chromosome 19.\n",
      "Adding map to HDF5...\n",
      "Intersection 18018 out of 18092 target HDF5 SNPs. 74 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch19.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch20.1240k.vcf -T data/filters/snps_bcftools_ch20.csv -M2 -v snps data/vcf.raw/example_hazelton_chr20.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Deleting previous HDF5 file at path_h5: test/example_hazelton.ch20.h5...\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 28940 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 30377 Chr.20 1240K SNPs.\n",
      "Intersection 28827 out of 28940 HDF5 SNPs\n",
      "Interpolating 113 variants.\n",
      "Finished Chromosome 20.\n",
      "Adding map to HDF5...\n",
      "Intersection 28827 out of 28940 target HDF5 SNPs. 113 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch20.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch21.1240k.vcf -T data/filters/snps_bcftools_ch21.csv -M2 -v snps data/vcf.raw/example_hazelton_chr21.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 15707 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 16727 Chr.21 1240K SNPs.\n",
      "Intersection 15640 out of 15707 HDF5 SNPs\n",
      "Interpolating 67 variants.\n",
      "Finished Chromosome 21.\n",
      "Adding map to HDF5...\n",
      "Intersection 15640 out of 15707 target HDF5 SNPs. 67 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch21.h5\n",
      "\n",
      "Print downsampling to 1240K...\n",
      "Running bash command: \n",
      "bcftools view -Ov -o test/example_hazelton.ch22.1240k.vcf -T data/filters/snps_bcftools_ch22.csv -M2 -v snps data/vcf.raw/example_hazelton_chr22.vcf.gz\n",
      "Finished BCF tools filtering to target markers.\n",
      "Converting to HDF5...\n",
      "Finished conversion to hdf5!\n",
      "Merging in LD Map..\n",
      "Lifting LD Map from eigenstrat to HDF5...\n",
      "Loaded 15483 variants.\n",
      "Loaded 6 individuals.\n",
      "Loaded 16420 Chr.22 1240K SNPs.\n",
      "Intersection 15408 out of 15483 HDF5 SNPs\n",
      "Interpolating 75 variants.\n",
      "Finished Chromosome 22.\n",
      "Adding map to HDF5...\n",
      "Intersection 15408 out of 15483 target HDF5 SNPs. 75 SNPs set to AF=0.5\n",
      "Transformation complete! Find new hdf5 file at: test/example_hazelton.ch22.h5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "map_path='./data/afs/v51.1_1240k.snp'\n",
    "\n",
    "for ch in {1..22};\n",
    "do\n",
    "    marker_path=data/filters/snps_bcftools_ch$ch.csv\n",
    "    af_path=data/afs/v51.1_1240k_AF_ch$ch.tsv\n",
    "    vcf_path=data/vcf.raw/example_hazelton_chr$ch.vcf.gz\n",
    "    ancIBD-run --vcf $vcf_path \\\n",
    "        --ch $ch --out test --marker_path $marker_path --map_path $map_path --af_path $af_path --prefix example_hazelton\n",
    "done"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\"> \n",
    "\n",
    "Note\n",
    "\n",
    "For large sample sizes, we recommend that one parallizes over autosomes for speed-up (e.g, by submitting array jobs on a cluster). The above for-loop is efficient only for small sample sizes. \n",
    "\n",
    "</div>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine IBD over 22 autosomes and generate summary statistics\n",
    "\n",
    "Now that we have individual IBD files for each of the autosome, we can combine the information across chromosomes and obtain genome-wide summary statistics for all pairs of samples (Only pairs of samples that share at least one IBD passing the length cutoff are recorded). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome 1; Loaded 10 IBD\n",
      "Chromosome 2; Loaded 9 IBD\n",
      "Chromosome 3; Loaded 6 IBD\n",
      "Chromosome 4; Loaded 9 IBD\n",
      "Chromosome 5; Loaded 8 IBD\n",
      "Chromosome 6; Loaded 7 IBD\n",
      "Chromosome 7; Loaded 9 IBD\n",
      "Chromosome 8; Loaded 7 IBD\n",
      "Chromosome 9; Loaded 6 IBD\n",
      "Chromosome 10; Loaded 7 IBD\n",
      "Chromosome 11; Loaded 5 IBD\n",
      "Chromosome 12; Loaded 5 IBD\n",
      "Chromosome 13; Loaded 8 IBD\n",
      "Chromosome 14; Loaded 6 IBD\n",
      "Chromosome 15; Loaded 3 IBD\n",
      "Chromosome 16; Loaded 6 IBD\n",
      "Chromosome 17; Loaded 4 IBD\n",
      "Chromosome 18; Loaded 5 IBD\n",
      "Chromosome 19; Loaded 8 IBD\n",
      "Chromosome 20; Loaded 6 IBD\n",
      "Chromosome 21; Loaded 6 IBD\n",
      "Chromosome 22; Loaded 6 IBD\n",
      "Saved 146 IBD to test/ch_all.tsv.\n",
      "> 8.0 cM: 146/146\n",
      "Of these with suff. SNPs per cM> 220:               113/146\n",
      "4     9\n",
      "2     8\n",
      "1     7\n",
      "13    7\n",
      "6     7\n",
      "8     7\n",
      "10    7\n",
      "21    6\n",
      "5     6\n",
      "7     6\n",
      "16    6\n",
      "11    5\n",
      "9     5\n",
      "12    4\n",
      "18    4\n",
      "20    4\n",
      "3     4\n",
      "14    3\n",
      "17    3\n",
      "22    3\n",
      "15    2\n",
      "Name: ch, dtype: int64\n",
      "Saved 9 individual IBD pairs to: test/ibd_ind.tsv\n"
     ]
    }
   ],
   "source": [
    "!ancIBD-summary --tsv test/example_hazelton.ch --out test/"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To view the complete options provided by the two command-line interface, use -h. For power users or people interested in applying the method beyond 1240k SNP set, keep in mind that one can obtain maximum flexibility by writing one's own wrappers (see section [prepare input](create_hdf5_from_vcf.ipynb), [run ancIBD](run_ancIBD.ipynb), and [visualization](plot_IBD.ipynb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: ancIBD-run [-h] [--vcf VCF] [--h5 H5] --ch CH --marker_path MARKER_PATH\n",
      "                  --map_path MAP_PATH [--af_path AF_PATH] [--out OUT]\n",
      "                  [--prefix PREFIX] [--min MIN] [--iid IID] [--pair PAIR]\n",
      "\n",
      "Run ancIBD.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --vcf VCF             path to the imputed vcf file\n",
      "  --h5 H5               path to hdf5 file. If specified, ancIBD will skip the\n",
      "                        vcf to hdf5 conversion step. Only one of --vcf and\n",
      "                        --h5 should be specified.\n",
      "  --ch CH               chromosome number (1-22).\n",
      "  --marker_path MARKER_PATH\n",
      "                        path to the marker file\n",
      "  --map_path MAP_PATH   path to the map file\n",
      "  --af_path AF_PATH     path to the allele frequency file (optional)\n",
      "  --out OUT             output folder to store IBD results and the\n",
      "                        intermediary .hdf5 file. If not specified, the results\n",
      "                        will be stored in the same folder as the input vcf\n",
      "                        file.\n",
      "  --prefix PREFIX       prefix of output file. If not specified, the prefix\n",
      "                        will be the same as the input vcf\n",
      "  --min MIN             minimum length of IBD segment in cM. Default is 8.\n",
      "  --iid IID             A list of sample iids to run ancIBD on (each line\n",
      "                        contains one sample IID). The sample list must match\n",
      "                        the sample name in the provided vcf file. If\n",
      "                        unspecified, ancIBD will run on all samples in the vcf\n",
      "                        file\n",
      "  --pair PAIR           A list of sample pairs to run ancIBD on (each line\n",
      "                        contains two sample IIDs separated by a whitespace).\n",
      "                        The sample list must match the sample name in the\n",
      "                        provided vcf file, and, if --iid is specified, all\n",
      "                        samples must also appear in the iid file. If\n",
      "                        unspecified, ancIBD will run on all pairs of samples\n",
      "                        in the vcf file\n"
     ]
    }
   ],
   "source": [
    "!ancIBD-run -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: ancIBD-summary [-h] --tsv TSV [--ch CH] [--bin BIN] [--snp_cm SNP_CM]\n",
      "                      [--out OUT]\n",
      "\n",
      "Run ancIBD.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help       show this help message and exit\n",
      "  --tsv TSV        base path to the individual IBD files.\n",
      "  --ch CH          chromosome number, expressed in the format chrom-chrom,\n",
      "                   e.g, 1-22). The default is 1-22.\n",
      "  --bin BIN        length bin over which IBD sharing summary statistics for\n",
      "                   pairs of samples will be calculated. Default is 8,12,16,20.\n",
      "  --snp_cm SNP_CM  minimum number of SNPs per centimorgan for a segment to be\n",
      "                   considered. The default is 220 to reduce false positive\n",
      "                   rates.\n",
      "  --out OUT        output folder to store results. If not specified, the\n",
      "                   results will be stored in the current directory.\n"
     ]
    }
   ],
   "source": [
    "!ancIBD-summary -h"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
